### PR TITLE
Fourier join fix

### PR DESCRIPTION
**Pull request by RainerHeintzmann**
I introduced a recursive type scheme. This seems to work now and the @report_opt seems to be happy, but there may be a cleaner or simpler way to achieve type stability.
Felix, can you do some speed test of the old and new scheme before merging?

### CODE REVIEW

**Comment by roflmaostc** [comment]
Thanks for that.

Did you check for type instabilities?
With JET?

**Comment by roflmaostc** [comment]
See here: https://github.com/bionanoimaging/FourierTools.jl/issues/29

File: src/custom_fourier_types.jl [hunk 1]
```diff
@@ -1,20 +1,22 @@
 ##  This View checks for the index to be L1 or the mirrored version (L2)
 # and then replaces the value by half of the parent at L1
+# do_split is a Bool that indicates whether this mechanism is active. It is needed for type stability reasons of functions returnting this type
```
Comment (roflmaostc) [inline src/custom_fourier_types.jl]: `do_split`.
Always wrap code in single ` `

File: src/custom_fourier_types.jl [hunk 2]
```diff
@@ -36,22 +38,24 @@ end
 
 ## This View checks for the index to be L1 
 # and then replaces the value by add the value at the mirrored position L2
+# do_join is a Bool that indicates whether this mechanism is active. It is needed for type stability reasons of functions returnting this type
```
Comment (roflmaostc) [inline src/custom_fourier_types.jl]: `do_join`

File: src/fourier_resizing.jl [hunk 3]
```diff
@@ -138,60 +138,85 @@ function rft_pad(mat, new_size)
...
-    for d = start_dim:ndims(mat)
-        sn = size_new[d]
-        so = size_old[d]
-        if sn < so && iseven(sn)
-            L1 = (size_old[d] -size_new[d] )÷2 +1
-            mat = FourierJoin(mat, d, L1)
-        end
+"""
+    ft_fix_before(mat::MT, size_old, size_new, ::Val{N})::FourierJoin{T,N,MT}  where {T,N, MT<:AbstractArray{T,N}} 
+    implements the specialized (highest dimension) version of a recursive dimension-specific function that returns an array type which knows
```
Comment (roflmaostc) [inline src/fourier_resizing.jl]: empty line after function definition

and move the to the very left:

So the text should not be tabbed to the right

File: src/fourier_resizing.jl [hunk 4]
```diff
@@ -138,60 +138,85 @@ function rft_pad(mat, new_size)
...
+function ft_fix_before(mat::MT, size_old, size_new, ::Val{N})::FourierJoin{T,N,MT}  where {T,N, MT<:AbstractArray{T,N}} 
+    sn = size_new[N]
+    so = size_old[N]
+    do_join = (sn < so && iseven(sn))
+    L1 = (size_old[N] -size_new[N] )÷2 +1
+    return FourierJoin(mat, N, L1, do_join)
+end
+
+"""
+    ft_fix_before(mat::MT, size_old, size_new, ::Val{D}=Val(1))  where {D, T, N, MT<:AbstractArray{T,N}}
```
Comment (roflmaostc) [inline src/fourier_resizing.jl:155]: same here

File: src/fourier_resizing.jl [hunk 5]
```diff
@@ -138,60 +138,85 @@ function rft_pad(mat, new_size)
...
-        if sn > so && iseven(so)
-            L1 = (size_new[d]-size_old[d])÷2+1
-            mat = FourierSplit(mat,d,L1)
-        end
-        # if equal do nothing
+function ft_fix_after(mat::MT, size_old, size_new, ::Val{N})::FourierSplit{T,N,MT}   where {T, N, MT<:AbstractArray{T,N}}
+    sn = size_new[N]
+    so = size_old[N]
+    do_split = (sn > so && iseven(so))
+    L1 = (size_new[N]-size_old[N])÷2+1
```
Comment (roflmaostc) [inline src/fourier_resizing.jl]: add spaces between operators

File: src/fourier_resizing.jl [hunk 6]
```diff
@@ -138,60 +138,85 @@ function rft_pad(mat, new_size)
...
 function rft_fix_first_dim_after(mat,size_old,size_new;dim=1)
     sn = size_new[dim] # Note that this dim is the corresponding real-space size
     so = size_old[dim] # Note that this dim is the corresponding real-space size
-    if sn > so && iseven(so) # source size is even upon padding
-        L1 = size_old[dim] ÷ 2 + 1
-        mat = FourierSplit(mat,dim,L1,-1) # This hack prevents a second position to be affected
-    end
+    do_split = (sn > so && iseven(so)) # source size is even upon padding
+    L1 = size_old[dim] ÷ 2 + 1
+    mat = FourierSplit(mat,dim,L1,-1, do_split) # This hack prevents a second position to be affected
```
Comment (roflmaostc) [inline src/fourier_resizing.jl]: space after a `, `

Comment (roflmaostc) [reply src/fourier_resizing.jl]: Comments always in front of a line. We don't do inline comments at the end

File: src/fourier_resizing.jl [hunk 7]
```diff
@@ -138,60 +138,85 @@ function rft_pad(mat, new_size)
...
     # if equal do nothing
     return mat
 end
 
 function rft_fix_before(mat,size_old,size_new)
     mat=rft_fix_first_dim_before(mat,size_old,size_new;dim=1) # ignore the first dimension
-    ft_fix_before(mat,size_old,size_new;start_dim=2) # ignore the first dimension
+    ft_fix_before(mat,size_old,size_new, Val(2)) # ignore the first dimension since it starts at Val(2)
```
Comment (roflmaostc) [inline src/fourier_resizing.jl]: fix comments

File: test/custom_fourier_types.jl [hunk 8]
```diff
@@ -2,10 +2,10 @@
 @testset "Custom Fourier Types" begin
     N = 5
     x = randn((N, N))
-    fs = FourierTools.FourierSplit(x, 2, 2, 4)
+    fs = FourierTools.FourierSplit(x, 2, 2, 4, true)
     @test FourierTools.parenttype(fs) == typeof(x)
     
-    fj = FourierTools.FourierJoin(x, 2, 2, 4)
+    fj = FourierTools.FourierJoin(x, 2, 2, 4, true)
```
Comment (roflmaostc) [inline test/custom_fourier_types.jl:10]: what about the `false` test case?

File: src/convolutions.jl [hunk 9]
```diff
@@ -3,7 +3,7 @@ export plan_conv_buffer, plan_conv_psf_buffer
 
 
 """
-    conv(u, v[, dims])
+conv(u, v[, dims])
```
Comment (roflmaostc) [inline src/convolutions.jl]: please revert all the tabbing changes.

The function has to be indented, the description not!
https://github.com/invenia/BlueStyle#documentation

**Comment by roflmaostc** [comment]
Please revert all the tabbing changes (with a git revert).
https://github.com/invenia/BlueStyle#documentation

File: src/custom_fourier_types.jl (hunk 1 above)
Comment (RainerHeintzmann) [reply src/custom_fourier_types.jl]: done

File: src/fourier_resizing.jl (hunk 3 above)
Comment (RainerHeintzmann) [reply src/fourier_resizing.jl]: done

File: src/fourier_resizing.jl (hunk 5 above)
Comment (RainerHeintzmann) [reply src/fourier_resizing.jl]: done

File: src/fourier_resizing.jl (hunk 7 above)
Comment (RainerHeintzmann) [reply src/fourier_resizing.jl]: done

File: test/custom_fourier_types.jl (hunk 8 above)
Comment (RainerHeintzmann) [reply test/custom_fourier_types.jl:10]: done

File: src/custom_fourier_types.jl [hunk 10]
```diff
@@ -1,20 +1,31 @@
-##  This View checks for the index to be L1 or the mirrored version (L2)
-# and then replaces the value by half of the parent at L1
+"""
+FourierSplit{T,N, AA<:AbstractArray{T, N}} <: AbstractArray{T,N}
```
Comment (roflmaostc) [inline src/custom_fourier_types.jl]: tab

File: src/custom_fourier_types.jl [hunk 11]
```diff
@@ -1,20 +1,31 @@
...
-    function FourierSplit(parent::AA, D::Int,L1::Int,L2::Int) where {T,N, AA<:AbstractArray{T, N}}
-        return new{T,N, AA}(parent, D, L1, L2)
+This version below is needed to avoid a split for the first rft dimension,
+but still return half the value FFTs and other RFT dimension should use the version without L2.
+"""
+    function FourierSplit(parent::AA, D::Int,L1::Int,L2::Int, do_split::Bool) where {T,N, AA<:AbstractArray{T, N}}
+        return new{T,N, AA}(parent, D, L1, L2, do_split)
     end
-    function FourierSplit(parent::AA, D::Int,L1::Int) where {T,N, AA<:AbstractArray{T, N}}
+    function FourierSplit(parent::AA, D::Int, L1::Int, do_split::Bool) where {T,N, AA<:AbstractArray{T, N}}
```
Comment (roflmaostc) [inline src/custom_fourier_types.jl:25]: no tab

File: src/custom_fourier_types.jl [hunk 12]
```diff
@@ -1,20 +1,31 @@
...
+    do_split::Bool
+
+"""
+    FourierSplit(parent::AA, D::Int,L1::Int,L2::Int, do_split::Bool) where {T,N, AA<:AbstractArray{T, N}}
 
-    # This version below is needed to avoid a split for the firs rft dimension but still return half the value
-    # FFTs and other RFT dimension should use the version without L2
-    function FourierSplit(parent::AA, D::Int,L1::Int,L2::Int) where {T,N, AA<:AbstractArray{T, N}}
-        return new{T,N, AA}(parent, D, L1, L2)
+This version below is needed to avoid a split for the first rft dimension,
```
Comment (roflmaostc) [inline src/custom_fourier_types.jl:19]: Tab the doc string in the case, I guess.

File: src/resampling.jl [hunk 13]
```diff
@@ -158,7 +158,7 @@ function upsample2(mat::AbstractArray{T, N}; dims=1:N, fix_center=false, keep_si
 end
 
 """
-    upsample2_abs2(mat::AbstractArray{T, N}; dims=1:N)
+upsample2_abs2(mat::AbstractArray{T, N}; dims=1:N)
```
Comment (roflmaostc) [inline src/resampling.jl]: tab

File: src/utils.jl [hunk 14]
```diff
@@ -169,7 +168,7 @@ end
 
 
 """
-    ft_center_diff(s [, dims])
+ft_center_diff(s [, dims])
```
Comment (roflmaostc) [inline src/utils.jl]: tab

File: src/custom_fourier_types.jl (hunk 2 above)
Comment (RainerHeintzmann) [reply src/custom_fourier_types.jl]: done

File: src/utils.jl (hunk 14 above)
Comment (RainerHeintzmann) [reply src/utils.jl]: done

File: src/custom_fourier_types.jl (hunk 12 above)
Comment (RainerHeintzmann) [reply src/custom_fourier_types.jl:19]: should be fine now

File: src/resampling.jl (hunk 13 above)
Comment (RainerHeintzmann) [reply src/resampling.jl]: done

File: src/custom_fourier_types.jl (hunk 10 above)
Comment (RainerHeintzmann) [reply src/custom_fourier_types.jl]: done

**Pull request merged**